Editor update: Map import, merge. Export .bsp models. Export .obj.

2020karaulov2020 opened the pull request:
Editor update: Map import, merge. Export .bsp models. Export .obj.

And some fixes.

2020karaulov2020 commented:
@wootguy I revert "merge" changes and now it works as in cmd line.

But merge map + model not works. Now model display in bspguy, but in game model invisibled ))))

@wootguy you can now check all changes and merge this pull request if all works ?)))

2020karaulov2020 commented:
@wootguy I added full .obj export (with textures). But some 3d viewers - crashed. But in 3d max I can see nice but mirrored by X or Z or Y, I don't know)))

2020karaulov2020 commented:
@wootguy import bsp model as entity ('lifehack' cycler_sprite + func_wall)

2020karaulov2020 commented:
@wootguy import bsp model as entity ('lifehack' cycler_sprite + func_wall)

[![ADD MODELS TO MAP](https://img.youtube.com/vi/8hExG9Cm3_w/0.jpg)](https://www.youtube.com/watch?v=8hExG9Cm3_w)

wootguy commented on src/editor/Gui.cpp:
```diff
@@ -222,6 +229,66 @@ void Gui::pasteLightmap() {
+	tmpMap->ents.push_back(tmpEnt);
+	tmpMap->update_ent_lump();
+
+	tmpMap->modelCount = 1;
+	tmpMap->models[0] = tmpModel;
```
I think this what is complicating the merge/import. The entity model is being exported as the worldspawn model here. The merger will try to combine this with the other map's worldspawn model, instead of appending to the list of models. Try deleting most of the worldspawn model structures and exporting the entity model as an entity model. You should be able to merge the maps without any code changes to BspMerger. After the maps are merged, then you can spawn an entity using the newly imported model index.

I think that would be cleaner than having the separate `MergeSecondsMapAsModel` mode in the merger. There might still need to be code to skip merging model 0, but it would make more sense.

2020karaulov2020 replied:
I got really works models at outpup (is valid and can be used in game). I removed MergeSecondsMapAsModel mode because it not works. )

2020karaulov2020 replied:
I tried also to add empty BSP model as worldspawn.  I see changes in bspguy, but in game I can't see these models. Now this option removed from bspguy.

wootguy commented on src/editor/Gui.cpp:
```diff
@@ -551,133 +637,256 @@ void ImportWad(Bsp* map, Renderer* app)
-			//map->write("yabma_move.bsp");
-			//map->write("D:/Steam/steamapps/common/Sven Co-op/svencoop_addon/maps/yabma_move.bsp");
-			map->write(map->path);
+		if (ImGui::BeginMenu("Save")) {
+			for (int r = 0; r < app->mapRenderers.size(); r++)
+			{
+				if (!app->mapRenderers[r]->map)
+					continue;
+				if (ImGui::MenuItem((std::string("Map ") + app->mapRenderers[r]->map->name + ".bsp").c_str())) {
+					Bsp* map = app->mapRenderers[r]->map;
+					if (map)
+					{
+						map->update_ent_lump();
+						map->update_lump_pointers();
+						map->write(map->path);
+					}
+				}
+			}
+			ImGui::EndMenu();
+		}
```
Let's keep this simple for now and not allow having multiple maps open at the same time. Copy+paste and some other parts of the UI probably need updating too, and this PR is already pretty big.

2020karaulov2020 replied:
But it needs to save models with needs coordinates. (It impossible without these option)

wootguy commented on src/editor/Gui.cpp:
```diff
@@ -915,6 +1124,9 @@ void Gui::drawMenuBar() {
 			showLightmapEditorWidget = !showLightmapEditorWidget;
 			showLightmapEditorUpdate = true;
 		}
+		if (ImGui::MenuItem("Map merge", "", showMergeMapWidget)) {
+			showMergeMapWidget = !showMergeMapWidget;
+		}
```
leave out for now

2020karaulov2020 replied:
I remove "model merge option" and now it works as in command line.

wootguy replied:
Not needed since there is no way to open multiple maps.

wootguy commented on src/editor/Gui.cpp:
```diff
@@ -2617,6 +2829,153 @@ void Gui::drawAbout() {
 	ImGui::End();
 }
 
+void Gui::drawMergeWindow() {
```
leave out

2020karaulov2020 replied:
Now it works as from command line. "For non linux users" )))

wootguy commented on src/editor/Gui.cpp:
```diff
@@ -2617,6 +2829,153 @@ void Gui::drawAbout() {
+						else
+						{
+							Bsp* map = g_app->mapRenderers[0]->map;
+							logf("Adding cycler_sprite at pos -8000 -8000 -8000 to precache model.\n");
```
why -8000? can it be created in front of the camera instead?

2020karaulov2020 replied:
> why -8000? can it be created in front of the camera instead?

Need to be hide it )) (Using only for precache .bsp model)

wootguy replied:
I realized later this is for a feature I don't think should be here. Using the "create entity" option and setting the model to a BSP file is good enough for now. I know you can't see the model that way, but this is confusing implementation for an `Open` option. Eventually the entity renderer will be updated to show the MDL/BSP instead of a colored cube, which would let you see the same thing without having to `Open` a map to see what a model looks like for a specific entity.

2020karaulov2020 replied:
Yes It used for adding bsp model to map. (Auto create cycler_sprite and func_wall with selected models, without 'create entity' button)

2020karaulov2020 replied:
I can use func_breakable for precache and show model with real geometry.

wootguy reviewed (changes requested):
So `Open` is adding a viewable model for the map in the editor, but not modifying the map? I was thinking the couch was merged into the map. It's cool to have that visualization, but I think the renderer should be updated instead to render BSP/MDL models automatically for entities, rather than having to go through this Open process for each one.

For `Open`, the map should be replaced with the opened map. I want to avoid showing multiple maps at the same time for now because there's a lot more that needs to be done for that to work properly.

Maybe we can start out simple and just have a PR for the model exporter? That seems to be working pretty well. I can review that then we can move on to the more complicated bits like model importing and multi-map editing.

2020karaulov2020 commented:
> So `Open` is adding a viewable model for the map in the editor, but not modifying the map? I was thinking the couch was merged into the map. It's cool to have that visualization, but I think the renderer should be updated instead to render BSP/MDL models automatically for entities, rather than having to go through this Open process for each one.
> 
> For `Open`, the map should be replaced with the opened map. I want to avoid showing multiple maps at the same time for now because there's a lot more that needs to be done for that to work properly.
> 

Okay.
I added 'Open' - to open single map. 'Add map' to open another maps to draw.

Removed model merger (not working).

Merge anyway not works ? I removed code modifications and it will be works as from command line.

2020karaulov2020 commented:
@wootguy I see "angles" value not supported by bspguy, and no rotation tool?

2020karaulov2020 commented:
@wootguy now instead cycler+wall I added import as 'func_breakable'. (Internal precache using **gibmodel**)

2020karaulov2020 commented:
@wootguy in latest commits :
1. Import entity as single entity : func_breakable (easy and no problems)
2. Fix entity report for imported models
3. Fix "Undo/Redo commands not updates draw entities"
4. Fixes crashes when works with more than one maps at same time.
5. 'Open' button works as open single map. 'Add' button - add another map to render. (For preview models, merge maps, export models, etc)
6. Removed 'merge model' option, now merge window works as in command line. 
7. etc

Now I fixed all problems ?

wootguy commented:
Overall, the code looks to work correctly. The issues I see are design-related.

I don't think the `Add Map` and `Import .bsp model as func_breakable` options are the right approach to rendering models for entities. I think it should work like it does in the Hammer/Jack editor where models are automatically loaded and used in place of the colored cube, if the model exists.

![entity_rendering](https://user-images.githubusercontent.com/12087544/111114391-667f5280-8520-11eb-90e2-900bcc1033d4.gif)
(like this but using a BSP file instead of agrunt.mdl)

The `Import` option for BSP models should merge the model into the BSP. Since that isn't working, I think it should be left out of this PR.

I don't think multi-map editing should be a part of this PR either (`Add Map` + `Save`/`Close` multiple files). I haven't decided what that should like yet, but I think showing both maps at once should be reserved for a special map-merging mode. Normal multi-map editing maybe should use tabs, or just allow copy-pasting between multiple bspguy windows.

2020karaulov2020 commented:
@wootguy

2020karaulov2020 commented:
@wootguy no alternatives to 'add map' at this time, I use it for model position, and it can be used to merge maps without command line(for 'non command line users'). Save button also used for save one model without original map, close button used for close single model without closing map. At first time it can be used as is. In future it can be replaced as you thinking.

@wootguy "angles" value in entity not supported by bspguy ?

2020karaulov2020 commented:
@wootguy in save setting function, flag 'ios::trunc' make empty file before writing. I do sometime lost settings when works with two or more copy of bspguy. Possible this flag is not needs ?

2020karaulov2020 commented:
@wootguy now to import model from one map to another needs do next steps:
1. Open first map
2. Open second map
3. Export one model from second map
4. Close second map
5. Open ('add map' button) model from second map
6. Place model at needed coordinates in first map
7. Save & close model
8. Press import as func_breakable
9. Save first map

How do you propose to simplify this?

wootguy commented:
> @wootguy no alternatives to 'add map' at this time, I use it for model position, and it can be used to merge maps without command line(for 'non command line users'). Save button also used for save one model without original map, close button used for close single model without closing map. At first time it can be used as is. In future it can be replaced as you thinking.

I'd rather not merge in code just to delete it later. I know there's no other way to visualize BSP models when used as an entity model, but I don't think this is the right approach to doing it. In the meantime there are server plugins that can adjust coordinates/angles in-game which can be then be copied to the entity properties.

> @wootguy "angles" value in entity not supported by bspguy ?

Right, it's ignored.

> @wootguy in save setting function, flag 'ios::trunc' make empty file before writing. I do sometime lost settings when works with two or more copy of bspguy. Possible this flag is not needs ? Or problem is not in flag?

Don't know, but it sounds like a race condition between both instances. There might need to be some sort of locking on the settings file to prevent both from modifying it at the same time.




> 
> 
> @wootguy now to import model from one map to another needs do next steps:
>   [...]
> How do you propose to simplify this?

The issue I have with importing isn't so much that it's complicated, but that it isn't really what I would call "Importing". The import function here is a fancy rendering mode for a single entity, shown temporarily to help you place that entity. I expect importing to merge that model into the BSP data, so that it could be used by its model index in brush entities.

That said, for what you're doing I think the process should be:
1. Export a model from Map A
1. Open map B
1. Set the model path of an entity to "maps/exported_model_from_map_b.bsp"
1. The editor automatically renders a BSP model in place of the colored cube, so you can position it easier

For what I would consider to be "Importing":
1. Export a model from Map A
1. Open map B
1. Import "maps/exported_model_from_map_b.bsp" (this would be a map merge which skips the worldspawn model)
1. Spawn a func_wall in front of the camera which uses the newly imported model index (e.g. "*123").

2020karaulov2020 commented:
> * Export a model from Map A
> * Open map B
> * Set the model path of an entity to "maps/exported_model_from_map_b.bsp"
> * The editor automatically renders a BSP model in place of the colored cube, so you can position it easier

But before it, you need save model with needed coordinates (because entity model not rendered in bspguy, you don't know what coordinates be in result)

I think that you are right and need to make the entity model display in bspguy. But there is no point in doing this only for bsp models, then need to add support for mdl ...

wootguy commented:
The point would be that the code isn't deleted in favor of something else later. MDL rendering wouldn't have to be completed at the same time, but I think that should eventually be implemented too.

> But before it, you need save model with needed coordinates (because entity model not rendered in bspguy, you don't know what coordinates be in result)

With entity model rendering implemented you'd be able to see the model after changing the model keyvalue. It might be far away from the entity origin though, if it's one of those models that has origin at (0,0,0). Maybe centering the model origin before exporting would be a useful option.

2020karaulov2020 commented:
> Maybe centering the model origin before exporting would be a useful option.

Yes it added in https://github.com/wootguy/bspguy/pull/41/commits/705638fab6de897f0a115d744496e470ddf9e599

2020karaulov2020 commented:
I tried to add tabs (windows) to open multiple maps without conflict.

https://github.com/wootguy/bspguy/pull/41/commits/a0af1d06bd328a556fa5f77f501ffcf3de5ef374

2020karaulov2020 commented:
@wootguy now I tried to add render bsp models from 'model' key. It works but now need do:

1. Open map
2. Open bsp model

Model attach to entity.

2020karaulov2020 commented:
@wootguy you can check all ?

Now added .bsp model auto load (if exists).

wootguy commented:
> I tried to add tabs (windows) to open multiple maps without conflict.
> 
> [a0af1d0](https://github.com/wootguy/bspguy/commit/a0af1d06bd328a556fa5f77f501ffcf3de5ef374)

Not bad, but shouldn't be part of this PR. I was thinking there should be labelled tabs at the top or bottom of the screen, but this menu saves some screen space which is nice.

> now I tried to add render bsp models from 'model' key. It works but now need do:

That's better, but manually opening a map in order to render a model for an entity in some other map is still really weird! The entity's model should be automatically loaded in the background or something, probably with a loading status since it could take while.

Again, I think this is PR is getting too complicated. Can we just start with the model exporting and review/finish that first? Ideally there would be one feature per PR.

Another reason this is taking longer is because a lot of this is design-related work that hasn't been decided yet (and I'm pretty picky about the UX). It could be helpful to plan out these features before coding. By agreeing how a feature should be implemented first, you wouldn't waste time coding something just to have me say "No! Not like that! 😱"

2020karaulov2020 commented:
@wootguy but in latest  commits I added "auto load bsp model" when open map. (search in assets path)

2020karaulov2020 commented:
@wootguy I can't separate model render and windows...

2020karaulov2020 commented:
@wootguy now all problems fixed???

wootguy commented:
The design issue still exists. Entity models should be loaded without having to open a new map which replaces the view of the map being edited. That said, it's a step closer now that the map loads automatically when you edit an entity model.

> @wootguy I can't separate model render and windows...  

That's ok, but I won't merge this code because I don't agree with the current approach. The model exporting code I think is mostly good to merge, but I haven't looked at the recent code changes because of the design issue.

2020karaulov2020 commented:
@wootguy If you open two map at the same time, bspguy hangs on 'In reloading'.  
I added wait for finish 'reloading'.
If you don't need to open multiple maps I can remove 'open' button.


Exported models marks as 'message:bsp model'. At map (.bsp model) opening, it saved to is_model variable.
If you open map with entity what used bsp model, bspguy automatically loaded this model as map (and mark as is_model), in render loop it find all models and attach to needed entity. 


@wootguy you can write what need redesign and I can try to do it)

wootguy commented:
> @wootguy If you open two map at the same time, bspguy hangs on 'In reloading'.  

Since BSP loading is slow, it might be best to load the primary map by itself first, then load all the other BSPs in a queue once the primary map finishes loading. That way the map which was opened is fully loaded and you can move around while the maps used as entity models are loaded in the background.

> If you don't need to open multiple maps I can remove 'open' button.  

It's not necessary for loading entity models, but it doesn't hurt to keep it either. It would be more useful to have it search for maps in the game directories, but that can come later.

> Exported models marks as 'message:bsp model'. At map (.bsp model) opening, it saved to is_model variable.  

This looks hacky. There can be a separate array of maps in the Renderer that are for use as entity models (e.g. `vector<BspRenderer*> entityMapModels;` in Renderer). The `mapRenderers` field is meant for multi-map editing, which is different from rendering entity models.

> @wootguy you can write what need redesign and I can try to do it)  

The GIF posted earlier is the goal:
https://github.com/wootguy/bspguy/pull/41#issuecomment-799168670

The map loading code likely needs to be refactored to allow loading maps in the background while a primary map is displayed. I'd offer more help but I'm still sort of taking a break from this project.

2020karaulov2020 commented:
> This looks hacky. There can be a separate array of maps in the Renderer that are for use as entity models (e.g. `vector<BspRenderer*> entityMapModels;` in Renderer). The `mapRenderers` field is meant for multi-map editing, which is different from rendering entity models.

I can use any bsp in 'Model' key as model ? Then at export it not need to be marked.



> > @wootguy you can write what need redesign and I can try to do it)
> 
> The GIF posted earlier is the goal:
> [#41 (comment)](https://github.com/wootguy/bspguy/pull/41#issuecomment-799168670)

At this moment (latest commit) it works as in GIF (in real time).

2020karaulov2020 commented:
Also in bspguy impossible to open two map at same time with shows only one map, need wait for load every map, map by map. (Before hide it in RenderLoop) I don't know why, and then I just added wait for load before access to load another map.

wootguy commented:
> I can use any bsp in 'Model' key as model ? Then at export it not need to be marked.

If I undertand how the BSP models work, then yes. I actually didn't know it was possible to use a BSP as an entity model until I saw your func_vehicle plugin. Is there something special about the BSPs used for entity models, or can you just use any map?

> At this moment (latest commit) it works as in GIF (in real time).  

When I edit the model key for an entity, the map is loaded automatically like in the GIF, but the map I was editing also disappears and is replaced with the new map. I also can't toggle the old map view from the `Windows` menu, but that's a separate issue.

> Also in bspguy impossible to open two map at same time with shows only one map, need wait for load every map, map by map. (Before hide it in RenderLoop) I don't know why, and then I just added wait for load before access to load another map.

yeah there's probably a good amount of refactoring needed for background loading to work. I did some refactoring to make viewing multiple maps possible, but I never thought that a BSP could be used as an entity model. It's a different way of thinking for loading map data, which probably means the code needs a lot of work to support that.

2020karaulov2020 commented:
@wootguy Yes BSP can be used as model (like as mdl) but need precache it before use. 

For this purpose possible to use cycler_sprite(but need move it outside world), or gibmodel parameter from func_breakable. (as in latest commits)

**You can use BSP model for bypass any .bsp limit (example convert part of map to func_wall and use it from another bsp file)**

2020karaulov2020 commented:
@wootguy In latest commit I fix problem with loading .bsp models (now can be used multiple models at one map)

2020karaulov2020 commented:
@wootguy video :
[![BSP PREVIEW](http://img.youtube.com/vi/K5-ryY_wMxw/0.jpg)](http://www.youtube.com/watch?v=K5-ryY_wMxw "BSP model drawing")

2020karaulov2020 commented:
@wootguy I tried to compile with new C++ but got conflict with 'byte' and 'std::byte' and 'byte'. In latest commit I fixed this conflict.

And fixed bad model export. (Create some data if not exists)

2020karaulov2020 commented:
In latest commit I added 'Load texture from .wad if not exists'. If you can write function for adding empty texture (only info) I can replace it.

wootguy commented:
The model rendering is looking good! Can you leave out the multi-map editing stuff for now? By that I mean the `Windows` menu, the submenu for `Save`, and for `Open` to replace the current map instead of adding a new one. I think the design for this isn't quite right.

When testing out the model feature, the BSP didn't go away when I deleted the `model` key (red X button). Also, if I toggle the model really fast the app crashes (backspace + letter + backspace + letter). I guess this happened because I edited the key before the map was finished loading. Preventing keyvalue edits while a map is loading is a good enough way to deal with that.

wootguy commented:
> You can use BSP model for bypass any .bsp limit (example convert part of map to func_wall and use it from another bsp file)

This is amazing and can help out a lot when merging maps that are near max clipnodes 😮 I need to try this out in Sven Co-op sometime.

2020karaulov2020 commented:
> > You can use BSP model for bypass any .bsp limit (example convert part of map to func_wall and use it from another bsp file)
> 
> This is amazing and can help out a lot when merging maps that are near max clipnodes 😮 I need to try this out in Sven Co-op sometime.

Yes but you can discover new problem, imported models caused error 'prediction error %f" (not world) , if player jump on this entity, onground flag is not active.

2020karaulov2020 commented:
> When testing out the model feature, the BSP didn't go away when I deleted the `model` key (red X button). Also, if I toggle the model really fast the app crashes (backspace + letter + backspace + letter). I guess this happened because I edited the key before the map was finished loading. Preventing keyvalue edits while a map is loading is a good enough way to deal with that.

Yes... it problem of asynchronous events.

2020karaulov2020 commented:
I don't know how to fix crash, I can just lock "model change" while model not loaded.
@wootguy you can write what I need to fix for merge this pull request?

wootguy commented:
> I don't know how to fix crash, I can just lock "model change" while model not loaded.  

Yeah that's fine. Disabling the text field or keyvalue window should work.

> @wootguy you can write what I need to fix for merge this pull request?
* Prevent the model-swapping crash
* Remove the multi-map editing stuff
  * Remove the `Windows` menu
  * Remove the submenu from `Save`
  * Have `Open` replace the current map instead of creating a new window (or just remove the option).
 
and then I'll take a look at the code after that.

2020karaulov2020 commented:
I want to update but I has no time, you can fix these problems and then merge it?))))

wootguy commented:
sure, might be a while though.

2020karaulov2020 commented:
> sure, might be a while though.

Can You can merge it as it is and then fix the problems?)))

I just can't find a one free hour to fix the latest problems 😭😭😭 (((

2020karaulov2020 commented:
> sure, might be a while though.

Can You can merge it as it is and then fix the problems?)))

I just can't find a one free hour to fix the latest problems  (((

wootguy commented:
That's the plan, but I don't want to spend on hour on this either yet. 😭 

I could do the fixes later, but then master isn't in a good state for others, possibly for months.

2020karaulov2020 commented:
😩😩😩😩

2020karaulov2020 commented:
@wootguy okay I remove "Windows" menu and now "Open" menu used as import (replace) current map.

wootguy commented:
nice, 2/4 done 

> * Prevent the model-swapping crash
> * Remove the multi-map editing stuff
>   * ~~Remove the `Windows` menu~~
>   * Remove the submenu from `Save`
>   * ~~Have `Open` replace the current map instead of creating a new window (or just remove the option).~~  
> 
> and then I'll take a look at the code after that.

2020karaulov2020 commented:
> nice, 2/4 done
> 
> > * Prevent the model-swapping crash
> > * Remove the multi-map editing stuff
> >   
> >   * ~Remove the `Windows` menu~
> >   * Remove the submenu from `Save`
> >   * ~Have `Open` replace the current map instead of creating a new window (or just remove the option).~
> > 
> > and then I'll take a look at the code after that.

but save no has submenu, it just write filename , do still need to delete?

wootguy commented:
> but save no has submenu,

This save submenu is what I mean. There can only ever be one map open so "Save" should just be a button.
![image](https://user-images.githubusercontent.com/12087544/126425708-35854e2f-c7f1-48d6-8d30-e17ade55e4b9.png)

> Prevent the model-swapping crash
> I forgot what it is, can you remind me?

Yeah it's in the earlier comments:
https://github.com/wootguy/bspguy/pull/41#issuecomment-809895303

2020karaulov2020 commented:
@wootguy okay i can fix it in few days, you can write that I can add more in these PR?

wootguy commented:
add more of what?

2020karaulov2020 commented:
No need more fixes?

wootguy commented:
Oh, no that's good. I thought you might mean adding new features.

2020karaulov2020 commented:
@wootguy now it fixed ?

2020karaulov2020 commented:
@wootguy @github-actions
C/C++ CI / build-linux (pull_request) Failing after 48s — build-linux

Need enable c++11 in linux build, i can't fix all c++11 code to previous version c++)))))

wootguy commented:
> 
> 
> @wootguy now it fixed ?

Looks good. I'll review the code next.

> Need enable c++11 in linux build, i can't fix all c++11 code to previous version c++)))))  

yeah don't worry about the CI stuff. I don't know why it doesn't pick up the c++11 flag in the cmake file.

wootguy commented on src/editor/Gui.cpp:
```diff
@@ -2607,20 +2946,183 @@ void Gui::drawAbout() {
+						}
+						else
+						{
+							Bsp* map = g_app->getSelectedMap()->map;
+							logf("Binding .bsp model to func_breakable.\n");
+							Entity* tmpEnt = new Entity("func_breakable");
+							tmpEnt->setOrAddKeyvalue("origin", model->ents[0]->getOrigin().toKeyvalueString());
+							tmpEnt->setOrAddKeyvalue("gibmodel", "models/" + basename(Path));
+							tmpEnt->setOrAddKeyvalue("model", "models/" + basename(Path));
+							tmpEnt->setOrAddKeyvalue("spawnflags", "1");
+							map->ents.push_back(tmpEnt);
+							map->update_ent_lump();
+							logf("Success! Now you needs to copy model to path: %s\n", ("models/" + basename(Path)).c_str());
+
+							if (app->pickInfo.map->GetBspRender())
+							{
+								app->pickInfo.map->GetBspRender()->preRenderEnts();
+								app->updateEntConnections();
+								app->updateEntConnectionPositions();
+							}
+							
+							app->reloadBspModels();
```
this doesn't seem to do anything besides add a func_breakable object with a model key set to a bsp file that doesn't exist. I think it's better to just document somewhere that it's possible to use BSPs as entity models.

wootguy commented on src/editor/Gui.cpp:
```diff
@@ -3740,6 +4258,36 @@ void Gui::drawTextureTool() {
 					}
 				}
+
+				if (!validTexture)
+				{
+					for (auto & s : mapRenderer->wads)
+					{
+						if (s->hasTexture(textureName))
+						{
+							WADTEX* wadTex = s->readTexture(textureName);
+							int lastMipSize = (wadTex->nWidth / 8) * (wadTex->nHeight / 8);
+
+							COLOR3* palette = (COLOR3*)(wadTex->data + wadTex->nOffsets[3] + lastMipSize + 2 - 40);
+							BYTE* src = wadTex->data;
+
+							COLOR3* imageData = new COLOR3[wadTex->nWidth * wadTex->nHeight];
+
+							int sz = wadTex->nWidth * wadTex->nHeight;
+
+							for (int k = 0; k < sz; k++) {
+								imageData[k] = palette[src[k]];
+							}
+							map->add_texture(textureName, (BYTE *)imageData, wadTex->nWidth, wadTex->nHeight);
+							delete[] imageData;
+							delete wadTex;
+							mapRenderer->reloadTextures();
+							textureChanged = true;
+							ImGui::End();
+							return;
+						}
+					}
+				}
```
I don't think selecting a missing texture should automatically create one. Why do this?

2020karaulov2020 replied:
@wootguy i think this code create texture in map if not found.

wootguy commented on src/editor/Renderer.cpp:
```diff
@@ -397,7 +401,13 @@ void Renderer::renderLoop() {
 		{
 			lastTitleTime = glfwGetTime( );
-			glfwSetWindowTitle(window, std::string(std::string("bspguy - ") + getMapContainingCamera()->map->path).c_str());
+			if (BspRenderer * tmpRend = getSelectedMap())
+			{
+				if (tmpRend->map && tmpRend->map->path.size())
+				{
+					glfwSetWindowTitle(window, std::string(std::string("bspguy - ") + tmpRend->map->path).c_str());
+				}
+			}
```
this only needs to be set on map loads

2020karaulov2020 replied:
@wootguy yes this code change title if map opened or changed

wootguy commented on src/editor/Renderer.cpp:
```diff
@@ -424,16 +434,57 @@ void Renderer::renderLoop() {
 				highlightEnt = pickInfo.entIdx;
 			}
-			mapRenderers[i]->render(highlightEnt, transformTarget == TRANSFORM_VERTEX, clipnodeRenderHull);
+			
+			if (curMap && curMap->ents.size() && !isLoading)
+			{
+				if (curMap->is_model)
+				{
+					for (int n = 0; n < mapRenderers.size(); n++)
+					{
+						if (n == i)
+							continue;
+
+						Bsp* anotherMap = mapRenderers[n]->map;
+						if (anotherMap && anotherMap->ents.size())
+						{
+							for (int s = 0; s < anotherMap->ents.size(); s++)
+							{
+								Entity* tmpEnt = anotherMap->ents[s];
+								if (tmpEnt->hasKey("model"))
+								{
+									if (!modelidskip.count(s))
+									{
+										if (basename(tmpEnt->keyvalues["model"]) == basename(curMap->path))
+										{
+											curMap->ents[0]->setOrAddKeyvalue("origin", tmpEnt->getOrigin().toKeyvalueString());
+											modelidskip.insert(s);
+											break;
+										}
+									}
+								}
+							}
+						}
+					}
+				}
+			}
```
This is way too much nesting. Please break some of this out into other functions or something.

Also what is the purpose of this?

2020karaulov2020 replied:
Yes it can be simplified.

I think this code change origin if model adding to map?

wootguy commented on src/editor/Renderer.cpp:
```diff
@@ -552,6 +603,18 @@ void Renderer::postLoadFgdsAndTextures() {
 	fgdFuture = async(launch::async, &Renderer::loadFgds, this);
 }
 
+void Renderer::clearMaps() {
+	for (int i = 0; i < mapRenderers.size(); i++) {
+		delete mapRenderers[i];
+	}
+	mapRenderers.clear();
+	pickInfo.valid = false;
+	clearUndoCommands();
+	clearRedoCommands();
+
+	logf("Cleared map list\n");
+}
```
not needed

2020karaulov2020 replied:
Need remove this function????????

wootguy commented on src/editor/Renderer.cpp:
```diff
@@ -1295,9 +1362,11 @@ void Renderer::pickObject() {
 	pickInfo.bestDist = FLT_MAX;
+
 	for (int i = 0; i < mapRenderers.size(); i++) {
+		mapRenderers[i]->preRenderEnts();
```
this is kind of expensive to do on every selection. Why is it needed?

2020karaulov2020 replied:
@wootguy @wootguy i don't know why this code added, many time elapsed from adding..., I can remove it ?

wootguy commented on src/editor/Renderer.cpp:
```diff
@@ -1331,26 +1400,21 @@ void Renderer::pickObject() {
+			bool select = true;
+			for (int i = 0; i < selectedFaces.size(); i++) {
+				if (selectedFaces[i] == pickInfo.faceIdx) {
+					select = false;
+					selectedFaces.erase(selectedFaces.begin() + i);
+					break;
 				}
+			}
 
-				mapRenderers[pickInfo.mapIdx]->highlightFace(pickInfo.faceIdx, select);
+			g_app->getSelectedMap()->highlightFace(pickInfo.faceIdx, select);
 
-				if (select)
-					selectedFaces.push_back(pickInfo.faceIdx);
-			}
+			if (select)
+				selectedFaces.push_back(pickInfo.faceIdx);
+			
 		}
```
There's no multi-map editing yet, but I still want this there in preparation for that.

2020karaulov2020 replied:
@wootguy i not see problem in these code

wootguy commented on src/editor/Renderer.cpp:
```diff
@@ -1529,6 +1593,41 @@ void Renderer::getPickRay(vec3& start, vec3& pickDir) {
 	pickDir = (start - cameraOrigin).normalize(1.0f);
 }
 
+BspRenderer* Renderer::getSelectedMap() {
+	for (int i = 0; i < mapRenderers.size(); i++)
+	{
+		BspRenderer* s = mapRenderers[i];
+		if (s->map && s->map->is_selected)
+		{
+			return s;
+		}
+	}
+	return mapRenderers.size() ? mapRenderers[0] : NULL;
+}
+
+int Renderer::getSelectedMapId() {
+	for (int i = 0; i < mapRenderers.size(); i++)
+	{
+		BspRenderer* s = mapRenderers[i];
+		if (s->map && s->map->is_selected)
+		{
+			return i;
+		}
+	}
+	return 0;
+}
+
+void Renderer::setSelectedMap(int id) {
+	for (int i = 0; i < mapRenderers.size(); i++)
+	{
+		BspRenderer* s = mapRenderers[i];
+		if (s->map)
+		{
+			mapRenderers[id]->map->is_selected = i == id;
+		}
+	}
+}
+
```
the `pickInfo` field can be used instead, once the `mapIdx` change is reverted.

2020karaulov2020 replied:
I just remove mapidx for easy access maps, and preparing for multiple map edit

wootguy commented on src/editor/Renderer.cpp:
```diff
@@ -1557,7 +1657,104 @@ void Renderer::setupView() {
+
+	std::vector<BspRenderer*> sorted_renders;
+
+	for (int i = 0; i < mapRenderers.size(); i++)
+	{
+		if (!mapRenderers[i]->map->is_model)
+		{
+			sorted_renders.push_back(mapRenderers[i]);
+		}
+		else
+		{
+			delete mapRenderers[i];
+		}
+	}
+
+	mapRenderers = sorted_renders;
```
I'm having a hard time following this. It looks like all map renderers are replaced with a list of renderers that are for models only. Wouldn't that delete the primary map renderer?

2020karaulov2020 replied:
@wootguy I think this code remove models from main render, many time elapsed since I edit code, I don't know what's happening if I do remove it...

wootguy commented on src/editor/Renderer.cpp:
```diff
@@ -1557,7 +1657,104 @@ void Renderer::setupView() {
+			if (entity->hasKey("model"))
+			{
+				std::string modelPath = entity->keyvalues["model"];
+				if (modelPath.find(".bsp") != std::string::npos)
```
should be case insensitive

wootguy commented on src/editor/Renderer.cpp:
```diff
@@ -1557,7 +1657,104 @@ void Renderer::setupView() {
+
+	tryPaths.insert(tryPaths.end(), g_settings.resPaths.begin(), g_settings.resPaths.end());
+
+	for (auto const& bsprend : mapRenderers)
+	{
+		for (auto const& entity : bsprend->map->ents)
```
should `is_model` map renderers be skipped?

2020karaulov2020 replied:
Models add to render list tail, it's needs for something but i don't know for what)))

wootguy commented on src/editor/Renderer.cpp:
```diff
@@ -1557,7 +1657,104 @@ void Renderer::setupView() {
+					for (int i = 0; i < tryPaths.size(); i++) {
+						string tryPath = tryPaths[i] + modelPath;
+						string tryPath_full = g_settings.gamedir + tryPaths[i] + modelPath;
+						if (fileExists(tryPath)) {
+							Bsp* tmpBsp = new Bsp(tryPath);
+							tmpBsp->is_model = true;
+							if (tmpBsp->valid)
+							{
+								BspRenderer* mapRenderer = new BspRenderer(tmpBsp, bspShader, fullBrightBspShader, colorShader, pointEntRenderer);
+								mapRenderers.push_back(mapRenderer);
+							}
+							break;
+						}
+						if (fileExists(tryPath_full)) {
+							Bsp* tmpBsp = new Bsp(tryPath_full);
+							tmpBsp->is_model = true;
+							if (tmpBsp->valid)
+							{
+								BspRenderer* mapRenderer = new BspRenderer(tmpBsp, bspShader, fullBrightBspShader, colorShader, pointEntRenderer);
+								mapRenderers.push_back(mapRenderer);
+							}
+							break;
+						}
```
too much duplicate code. You can do something like this instead:
```
string existingPath = "";
if (fileExists(tryPath)) {
    existingPath = tryPath;
}
if (fileExists(tryPath_full)) {
    existingPath = tryPath_full;
}
if (existingPath.Length() > 0) {
   Bsp stuff
}
```

wootguy commented on src/editor/Renderer.cpp:
```diff
@@ -1557,7 +1657,104 @@ void Renderer::setupView() {
+}
+
 void Renderer::addMap(Bsp* map) {
+	if (!map->is_model)
+	{
+		for (int i = 0; i < mapRenderers.size(); i++)
+		{
+			mapRenderers[i]->map->is_selected = false;
+		}
+		map->is_selected = true;
+	}
```
just noticing this `is_selected` code. I don't think that needs to exist. The `pickInfo` var indicates which map is selected.

2020karaulov2020 replied:
I replaced pickinfo because it not easy to access maps

wootguy commented on src/editor/Renderer.cpp:
```diff
@@ -2897,6 +3094,12 @@ void Renderer::undo() {
 	undoCommand->undo();
 	undoHistory.pop_back();
 	redoHistory.push_back(undoCommand);
+	if (pickInfo.map->GetBspRender())
+	{
+		pickInfo.map->GetBspRender()->preRenderEnts();
+		updateEntConnections();
+		updateEntConnectionPositions();
+	}
```
why? Selecting faces shouldn't affect entity renderers

wootguy commented on src/editor/Renderer.cpp:
```diff
@@ -2913,6 +3116,12 @@ void Renderer::redo() {
 	redoCommand->execute();
 	redoHistory.pop_back();
 	undoHistory.push_back(redoCommand);
+	if (pickInfo.map->GetBspRender())
+	{
+		pickInfo.map->GetBspRender()->preRenderEnts();
+		updateEntConnections();
+		updateEntConnectionPositions();
+	}
```
why is this needed? I think it should be part of a specific undo command if it's required at all.

2020karaulov2020 replied:
@wootguy I think this command update ents position ??

2020karaulov2020 replied:
@wootguy okay, I move this code to new function, and replaced all matches

wootguy commented on src/main.cpp:
```diff
@@ -94,7 +94,7 @@
 
 
-const char* g_version_string = "bspguy v4 WIP (November 2020)";
+const char* g_version_string = "bspguy v4 WIP (March, 2021)";
```
probably going to be August

2020karaulov2020 replied:
@wootguy ok

wootguy commented on src/main.cpp:
```diff
@@ -669,51 +691,57 @@ int main(int argc, char* argv[])
 
-	if (argc == 2) {
-		start_viewer(argv[1]);
+	if (cli.command == "editor" || cli.command == "empty") {
+		start_viewer(std::string());
+		return 0;
 	}
```
how is this useful?

2020karaulov2020 replied:
For opening map from editor instead of command line

wootguy commented on src/main.cpp:
```diff
@@ -669,51 +691,57 @@ int main(int argc, char* argv[])
-		else {
-			logf("unrecognized command: %d\n", cli.command.c_str());
-		}
+	if (cli.command == "exportobj") {
+		Bsp* tmpBsp = new Bsp(cli.bspfile);
+		tmpBsp->ExportToObjWIP(cli.bspfile);
+		delete tmpBsp;
+		return 0;
+	}
```
this can move down with the other commands

2020karaulov2020 replied:
Move down? How i can understand it ?

wootguy commented on src/util/util.cpp:
```diff
@@ -694,6 +694,65 @@ bool pointInsidePolygon(vector<vec2>& poly, vec2 p) {
+#define MAX_NUMBER_OF_COLORS 0
+#define ALL_COLORS_REQUIRED 0
+
+void WriteImage(std::string fileName, BYTE* pixels, int width, int height, int bytesPerPixel)
```
Can rename to the type of image? I'm guessing `WriteBMP`

2020karaulov2020 replied:
Fixed

wootguy reviewed (changes requested):
Just noticed now that Entity Report widget is crashing the program when you select an entity. To reproduce:
1. Open the Entity report widget
1. Deslect the map (click in the black void)
1. Select an entity in the widget.

2020karaulov2020 commented on src/bsp/Bsp.cpp:
```diff
@@ -20,28 +24,29 @@ vec3 default_hull_extents[MAX_MAP_HULLS] = {
 		if (i == LUMP_TEXTURES)
 		{
-			lumps[i] = new byte[4];
+			lumps[i] = new BYTE[4];
```
I don't know why. Now I replace BYTE to byte.

2020karaulov2020 commented on src/editor/BspRenderer.h:
```diff
@@ -83,7 +83,7 @@ struct RenderClipnodes {
 
 struct PickInfo {
-	int mapIdx;
+	//int mapIdx;
```
I just extended functional of this variable, all moved to getSelectedMapId( ) and setSelectedMap() getSelectedMap()

2020karaulov2020 replied:
@wootguy I think it better that mapIdx in PickInfo.

2020karaulov2020 replied:
@wootguy moreover I want to remove Pick Info to adding more features (select multiple faces/models and etc)

2020karaulov2020 commented on src/editor/Gui.cpp:
```diff
@@ -219,7 +228,113 @@ void Gui::pasteLightmap() {
+	tmpMap->modelCount = 1;
+	tmpMap->models[0] = tmpModel;
+
+	// Move model to 0 0 0
+	tmpMap->move(-modelOrigin, 0, true);
```
@wootguy move model without origin.

2020karaulov2020 commented on src/bsp/Bsp.cpp:
```diff
@@ -506,7 +553,7 @@ bool Bsp::vertex_manipulation_sync(int modelIdx, vector<TransformVert>& hullVert
 }
 
-bool Bsp::move(vec3 offset, int modelIdx) {
+bool Bsp::move(vec3 offset, int modelIdx, bool onlyModel) {
```
@wootguy move model without origin.

2020karaulov2020 commented on src/cli/CommandLine.cpp:
```diff
@@ -38,6 +39,12 @@ CommandLine::CommandLine(int argc, char* argv[]) {
 		}
 	}
+
+	if (argc == 2)
+	{
+		bspfile = argv[1];
+		return;
+	}
```
@wootguy you can remove it and test ))

@wootguy this is fix for no bspfile in some cases.

2020karaulov2020 commented on src/editor/BspRenderer.cpp:
```diff
@@ -1345,6 +1353,10 @@ void BspRenderer::render(int highlightEnt, bool highlightAlwaysOnTop, int clipno
 }
 
 void BspRenderer::drawModel(int modelIdx, bool transparent, bool highlight, bool edgesOnly) {
+	if (modelIdx >= numRenderModels)
```
@wootguy fixed

2020karaulov2020 commented on src/editor/Gui.cpp:
```diff
@@ -219,7 +228,113 @@ void Gui::pasteLightmap() {
+
+	tmpMap->lumps[LUMP_MODELS] = (BYTE*)tmpMap->models;
+	tmpMap->header.lump[LUMP_MODELS].nLength = sizeof(tmpModel);
+	tmpMap->update_lump_pointers();
```
@wootguy okay, replaced.

2020karaulov2020 replied:
@wootguy I tried but it cause crash when model export.

2020karaulov2020 commented on src/editor/Gui.cpp:
```diff
@@ -219,7 +228,113 @@ void Gui::pasteLightmap() {
+	if (!removed.allZero())
+		removed.print_delete_stats(1);
+
+	if (!tmpMap->validate())
+	{
+		int markid = 0;
+		for (int i = 0; i < tmpMap->leafCount; i++)
+		{
+			BSPLEAF& tmpLeaf = tmpMap->leaves[i];
+			tmpLeaf.iFirstMarkSurface = markid;
+			markid += tmpLeaf.nMarkSurfaces;
+		}
+
+		while (tmpMap->models[0].nVisLeafs >= tmpMap->leafCount)
+			tmpMap->create_leaf(-2);
+	}
```
@wootguy I don't know why, but some models can't be validated because no leaf count matches.
@wootguy You can tell me how to fix it without these code ?

2020karaulov2020 commented on src/editor/Gui.cpp:
```diff
@@ -219,7 +228,113 @@ void Gui::pasteLightmap() {
+	*/
+
+	tmpMap->lumps[LUMP_LEAVES] = (BYTE*)tmpMap->leaves;
+	tmpMap->update_lump_pointers();
```
@wootguy okay it moved to **if (!tmpMap->validate())** block.

2020karaulov2020 commented on src/editor/Gui.cpp:
```diff
@@ -219,7 +228,113 @@ void Gui::pasteLightmap() {
+	tmpMap->header.lump[LUMP_MODELS].nLength = sizeof(tmpModel);
+	tmpMap->update_lump_pointers();
+
+	/*for (int i = 0; i < tmpMap->leafCount; i++)
+	{
+		BSPLEAF& tmpLeaf = tmpMap->leaves[i];
+		tmpLeaf.nVisOffset = -1; 
+		for (int n = 0; n < 3; n++)
+		{
+			tmpLeaf.nMins[n] = tmpLeaf.nMins[n] < tmpModel.nMins[n] ? tmpModel.nMins[n] : tmpLeaf.nMins[n];
+			tmpLeaf.nMaxs[n] = tmpLeaf.nMaxs[n] > tmpModel.nMaxs[n] ? tmpModel.nMaxs[n] : tmpLeaf.nMaxs[n];
+		}
+	}*/
```
@wootguy ok

2020karaulov2020 commented on src/editor/Gui.cpp:
```diff
@@ -219,7 +228,113 @@ void Gui::pasteLightmap() {
+		while (tmpMap->models[0].nVisLeafs >= tmpMap->leafCount)
+			tmpMap->create_leaf(-2);
+	}
+	/*
+	if (tmpMap->models[0].iHeadnodes[0] < 0)
+		tmpMap->regenerate_clipnodes(0, 0);
+	if (tmpMap->models[0].iHeadnodes[1] < 0)
+		tmpMap->regenerate_clipnodes(0, 1);
+	if (tmpMap->models[0].iHeadnodes[2] < 0)
+		tmpMap->regenerate_clipnodes(0, 2);
+	if (tmpMap->models[0].iHeadnodes[3] < 0)
+		tmpMap->regenerate_clipnodes(0, 3);
+	*/
```
@wootguy ok

2020karaulov2020 commented on src/editor/Gui.cpp:
```diff
@@ -551,133 +685,257 @@ void ImportWad(Bsp* map, Renderer* app)
+			BspRenderer* mapRenderer = app->getSelectedMap();
+			if (mapRenderer)
+			{
+				Bsp* map = mapRenderer->map;
+				if (map)
+				{
+					delete map;
+					mapRenderer->map = NULL;
+					std::vector<BspRenderer*> renders;
+					for (int r = 0; r < app->mapRenderers.size(); r++)
 					{
-						std::string entities = std::string(map->lumps[LUMP_ENTITIES], map->lumps[LUMP_ENTITIES] + map->header.lump[LUMP_ENTITIES].nLength - 1);
-						entFile.write(entities.c_str(), entities.size());
+						if (app->mapRenderers[r]->map)
+							renders.push_back(app->mapRenderers[r]);
+						else
+							delete app->mapRenderers[r];
 					}
+					app->mapRenderers = renders;
+					app->pickInfo = PickInfo();
 				}
 			}
-			if (ImGui::MenuItem("Embedded textures (.wad)", NULL)) {
-				Bsp* map = app->getMapContainingCamera()->map;
-				if (map)
+		}
```
@wootguy how to fix it ? (close map without close bspguy)

2020karaulov2020 commented on src/editor/Gui.cpp:
```diff
@@ -915,12 +1173,56 @@ void Gui::drawMenuBar() {
+	{
+		bool found_model = false;
+		for (int r = 0; r < app->mapRenderers.size(); r++)
+		{
+			if (!app->mapRenderers[r]->map)
+				continue;
+			if (app->mapRenderers[r]->map->is_model)
+			{
+				found_model = true;
+				continue;
+			}
+			if (ImGui::MenuItem((std::string("Map ") + app->mapRenderers[r]->map->name + ".bsp").c_str(), 0, r == app->getSelectedMapId()))
+			{
+				app->setSelectedMap(r);
+				app->pickInfo = PickInfo();
+			}
+		}
+
+		if (found_model && ImGui::BeginMenu("Models"))
+		{
+			for (int r = 0; r < app->mapRenderers.size(); r++)
+			{
+				if (!app->mapRenderers[r]->map)
+					continue;
+				if (!app->mapRenderers[r]->map->is_model)
+					continue;
+				if (ImGui::MenuItem((std::string("Model ") + app->mapRenderers[r]->map->name + ".bsp").c_str(), 0, r == app->getSelectedMapId()))
+				{
+					app->setSelectedMap(r);
+					app->pickInfo = PickInfo();
+				}
+			}
+
+			ImGui::EndMenu();
+		}
+
+		ImGui::EndMenu();
+	}
+	*/
+
```
@wootguy ok

2020karaulov2020 commented on src/editor/Gui.cpp:
```diff
@@ -1400,6 +1703,27 @@ void Gui::drawKeyvalueEditor_SmartEditTab(Entity* ent) {
 		static InputData inputData[128];
 		static int lastPickCount = 0;
 
+		if (ent->hasKey("model"))
+		{
+			bool foundmodel = false;
+			for (int i = 0; i < fgdClass->keyvalues.size(); i++) {
+				KeyvalueDef& keyvalue = fgdClass->keyvalues[i];
+				string key = keyvalue.name;
+				if (key == "model") 
+				{
+					foundmodel = true;
+				}
+			}
+			if (!foundmodel)
+			{
+				KeyvalueDef keyvalue = KeyvalueDef();
+				keyvalue.name = "model";
+				keyvalue.description = "Model";
+				keyvalue.iType = FGD_KEY_STRING;
+				fgdClass->keyvalues.push_back(keyvalue);
+			}
+		}
```
@wootguy it just marks bsp file as 'MODEL' (for bspguy)

2020karaulov2020 commented:
> Just noticed now that Entity Report widget is crashing the program when you select an entity. To reproduce:
> 
> 1. Open the Entity report widget
> 2. Deslect the map (click in the black void)
> 3. Select an entity in the widget.

Okay i can test and fix it in two days

2020karaulov2020 commented:
@wootguy I added support bspguy.cfg in working dir(crossplatform), if 'WorkingDir/bspguy.cfg' exists. Or no access to config file.

It no affect your code if no file created. 
you will not mind? :(

And fix error saving configs/etc.


`* out is always set for ofstream objects (even if explicitly not set in argument mode).` http://www.cplusplus.com/reference/fstream/ofstream/open/

2020karaulov2020 commented:
@wootguy `Entity Report widget is crashing the program `

Fixed

2020karaulov2020 commented:
@wootguy and now it compiled for linux. But without <filesystem>

2020karaulov2020 commented:
@wootguy I added fix crash in bsp, if you tried to move face in 'tranform' widget

2020karaulov2020 commented:
@wootguy you can merge it and then fix non critical problems ?

2020karaulov2020 commented:
@wootguy okay now I remove is_selected (multiple maps edition stuffs) and return back pickInfo.map

2020karaulov2020 commented:
@wootguy  I think pickInfo.valid is a parasitic function and is not needed at all. Valid entity? Valid map? Valid model ? what is .valid ?? I just remove it from all code. I think it better solution.

2020karaulov2020 commented:
@wootguy drag axes sometime shows with delay, even in original bspguy. (Sometime visible sometime not visibled)

2020karaulov2020 commented:
@wootguy many functions works bad with FLT_MIN FLT_MAX, because FLT_MIN > -9999.f and save with FLT_MAX < 9999.f....
It looks like strange c++ issue
For example scale axes shows at random position.

2020karaulov2020 commented:
@wootguy It really problem in FLT_MIN/MAX. I redefine it and attach to MAX_MAP_COORD.

2020karaulov2020 commented:
@wootguy now scale axes works. With true position. But one of scale axes works strange, it move object like move axes))))

2020karaulov2020 commented:
@wootguy now scale axes works well!

2020karaulov2020 commented:
@wootguy now scale axes works well!

2020karaulov2020 commented:
@wootguy how to move origin? I not see drag axes))) Even in original bspguy. I tried to fix problem

2020karaulov2020 commented:
@wootguy You can check all now ?

wootguy commented:
I'm trying to work up up the motivation to go through this all again. Even just reading/responding to changes takes hours. 😅 

> @wootguy you can merge it and then fix non critical problems ? 

I could do that, but I don't want to change/remove stuff behind your back either. It seems like we've got different ideas on how some things should work.

Another thought is that I'm limiting you on progress for this. I'm starting to think I'll never come back to this project, so it might be more efficient if you take the lead on development and start releasing builds in your branch. Then maybe later I can incorporate some of those changes on my own, if I ever get back to this.

2020karaulov2020 commented:
@wootguy but I can't do it, I don't have time to do it either) You can review new updates? I returned back pickinfo.map (And added bugfixes for old review). No updates more added.

wootguy commented:
yeah I'll take a look eventually. Likely this week or next.

2020karaulov2020 commented:
> yeah I'll take a look eventually. Likely this week or next.

I removed pickInfo.valid because this is not true. Just still check by pickinfo. 
And I found problem with FLT_MAX/FLT_MIN (yes, again). Problems with scaleaxes, etc. I replaced it with map coords max, and now it works, you can test autobuild with default FLT_MAX and new.

2020karaulov2020 commented:
> yeah I'll take a look eventually. Likely this week or next.
@wootguy okay , You can just write what bothers you, I will redo it if it is possible, to finally merge this pull request))

2020karaulov2020 commented:
@wootguy I found, **FLT_MIN not is FLT_MIN )**
` /* Minimum normalized positive floating-point number, b**(emin - 1).  */`
I think it more like to 0.0000000000000001 , but not to -999999999999

**Everyone who has ever used FLT_MIN has made a big mistake.**

🤣 🤣 🤣 🤣 🤣 🤣 🤣

2020karaulov2020 commented:
@wootguy I test compile with latest C++/C language. Using 'namespace std' cause problems with compilation and I removed it. Also added some fixes in code. (Fix some problems found by microsoft static analyzer)

2020karaulov2020 commented:
And now fixed linux build.

2020karaulov2020 commented:
@wootguy I test and think what all functions works properly.
But I can't test map merge, you can do it ?(No sven co op installed)

2020karaulov2020 commented:
@wootguy how I can add MDL renderer ? I add headers / sources. But I don't know how to use) 

I don't understand well in 3D, can you tell me how to draw a model somewhere on the map?

2020karaulov2020 commented:
@wootguy I added main_mdlviewer.cpp ( for test MDL VIEWER, need replace main.cpp with main_mdlviewer.cpp) and it works. 

But I don't know how to draw models in BSPGUY :)

2020karaulov2020 commented:
> Overall, the code looks to work correctly. The issues I see are design related.
> 
> I don't think the `Add Map` and `Import .bsp model as func_breakable` options are the right approach to rendering models for entities. I think it should work like it does in the Hammer/Jack editor where models are automatically loaded and used in place of the colored cube, if the model exists.
> 
> ![entity_rendering](https://user-images.githubusercontent.com/12087544/111114391-667f5280-8520-11eb-90e2-900bcc1033d4.gif)
> (like this but using a BSP file instead of agrunt.mdl)

Now I want to add support of .mdl files, but I'm not good at opengl, if you help me (at least an example of drawing a single model in bspguy coordinates), then I can do all the rest of the work myself.

2020karaulov2020 commented:
@wootguy or code from main_mdlviewer.cpp cannot be used in bspguy?
